nemo/collections/nlp/models/language_modeling/transformer_lm_model.py

```diff
@@ -0,0 +1,199 @@
+            attn_score_dropout=cfg.language_model.get("attn_score_dropout", 0.0),
+            attn_layer_dropout=cfg.language_model.get("attn_layer_dropout", 0.0),
+        )
+        self.log_softmax = TokenClassifier(
+            hidden_size=cfg.language_model.hidden_size, num_classes=vocab_size, log_softmax=True,
+        )
+
+        std_init_range = 1 / math.sqrt(cfg.language_model.hidden_size)
+        self.apply(lambda module: transformer_weights_init(module, std_init_range))
+
+        # tie weights of embedding and softmax matrices
+        self.log_softmax.mlp.layer0.weight = self.embedding_layer.token_embedding.weight
+
+        self.training_loss = SmoothedCrossEntropyLoss(pad_id=self.tokenizer.pad_id)
+        self.validation_loss = SmoothedCrossEntropyLoss(pad_id=self.tokenizer.pad_id, predict_last_k=64)
+
+        # Optimizer setup needs to happen after all model weights are ready
+        self.setup_optimization(cfg.optim)
+
+    @typecheck()
+    def forward(self, input_ids, attention_mask):
+        """
+        No special modification required for Lightning, define it as you normally would
+        in the `nn.Module` in vanilla PyTorch.
+        """
+        token_embeddings = self.embedding_layer(input_ids)
+        hidden_states = self.encoder(token_embeddings, attention_mask)
+        log_probs = self.log_softmax(hidden_states=hidden_states)
+
+        return log_probs
+
+    def training_step(self, batch, batch_idx):
+        """
+        Lightning calls this inside the training loop with the data from the training dataloader
+        passed in as `batch`.
+        """
+        # forward pass
+        input_ids, input_mask, labels = batch
+        log_probs = self(input_ids=input_ids, attention_mask=input_mask)
+
+        train_loss = self.training_loss(logits=log_probs, labels=labels)
+
+        tensorboard_logs = {'train_loss': train_loss, 'lr': self._optimizer.param_groups[0]['lr']}
+        return {'loss': train_loss, 'log': tensorboard_logs}
+
+    def validation_step(self, batch, batch_idx):
+        """
+        Lightning calls this inside the validation loop with the data from the validation dataloader
+        passed in as `batch`.
+        """
+        input_ids, input_mask, labels = batch
+        log_probs = self(input_ids=input_ids, attention_mask=input_mask)
+
+        val_loss = self.validation_loss(logits=log_probs, labels=labels)
+
+        tensorboard_logs = {
+            'val_loss': val_loss,
+        }
+
+        return {'val_loss': val_loss, 'log': tensorboard_logs}
+
+    def validation_epoch_end(self, outputs):
+        """
+        Called at the end of validation to aggregate outputs.
+        :param outputs: list of individual outputs of each validation step.
+        """
+
+        avg_loss = torch.stack([x['val_loss'] for x in outputs]).mean()
+        tensorboard_logs = {'val_loss': avg_loss, 'val_ppl': torch.exp(avg_loss)}
+        return {'val_loss': avg_loss, 'log': tensorboard_logs}
```

Comment:
do yo have any other eval metrics than the loss itself. e.g. Perplexity like here https://github.com/NVIDIA/NeMo/blob/candidate/nemo/collections/nlp/models/language_modeling/lm_model.py#L124

Reply:
Perplexity is saved to tensorboard logs and I am not sure if we need to add it to metrics dict too. All other models I looked at, stored val_loss only, other metrics were deferred to tensorboard logs.